http-service-lambda/src/lib.rs

```diff
@@ -0,0 +1,179 @@
+//! `HttpService` server that uses AWS Lambda Rust Runtime as backend.
+//!
+//! This crate builds on the standard http interface provided by the
+//! [lambda_http](https://docs.rs/lambda_http) crate and provides a http server
+//! that runs on the lambda runtime.
+//!
+//! Compatible services like [tide](https://github.com/rustasync/tide) apps can
+//! run on lambda and processing events from API Gateway or ALB without much
+//! change.
+//!
+//! # Examples
+//!
+//! **Hello World**
+//!
+//! ```rust,ignore
+//! #![feature(async_await)]
+//!
+//! fn main() {
+//!     let mut app = tide::App::new();
+//!     app.at("/").get(async move |_| "Hello, world!");
+//!     http_service_lambda::run(app.into_http_service());
+//! }
+//! ```
+
+#![forbid(future_incompatible, rust_2018_idioms)]
+#![deny(missing_debug_implementations, nonstandard_style)]
+#![warn(missing_docs, missing_doc_code_examples)]
+#![cfg_attr(test, deny(warnings))]
+#![feature(async_await)]
+
+use futures::{FutureExt, TryFutureExt};
+use http_service::{Body as HttpBody, HttpService, Request as HttpRequest};
+use lambda_http::{Body as LambdaBody, Handler, Request as LambdaHttpRequest};
+use lambda_runtime::{error::HandlerError, Context};
+use std::future::Future;
+use std::sync::Arc;
+use tokio::runtime::Runtime as TokioRuntime;
+
+type LambdaResponse = lambda_http::Response<lambda_http::Body>;
+
+trait ResultExt<OK, ERR> {
+    fn handler_error(self, description: &str) -> Result<OK, HandlerError>;
+}
+
+impl<OK, ERR> ResultExt<OK, ERR> for Result<OK, ERR> {
+    fn handler_error(self, description: &str) -> Result<OK, HandlerError> {
+        self.map_err(|_| HandlerError::from(description))
+    }
+}
+
+trait CompatHttpBodayAsLambda {
+    fn into_lambda(self) -> LambdaBody;
+}
+
+impl CompatHttpBodayAsLambda for Vec<u8> {
+    fn into_lambda(self) -> LambdaBody {
+        if self.is_empty() {
+            return LambdaBody::Empty;
+        }
+        match String::from_utf8(self) {
```

Comment:
This makes me uncomfortable, what if it is binary data which just happens to be utf-8 compatible? Do the different body types affect how the API gateway returns the bodies to the client? Or does it just affect the base64 encoding as it passes the value from here to the gateway?

Reply:
This is a good question. For API gateway, it doesn't matter very much because the API gateway will do the final conversion: https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-payload-encodings-workflow.html so the check here only save us one extra serialization to base64 and back.

For ELB, a base64 encoded body is required to return binary data: https://docs.aws.amazon.com/elasticloadbalancing/latest/application/lambda-functions.html#respond-to-load-balancer, but I think in the end it's the `content-type` header decides how the client should handle the response.

Technically we can always return base64 encoded data, but we can save the serialization effort when we don't need to.